peps/pep-0667.rst

```diff
@@ -40,72 +62,227 @@ The current implementation of ``locals()`` and ``frame.f_locals`` is slow,
 
-For example::
+For example, when attempting to manipulate local variables via frame objects::
 
     class C:
         x = 1
         sys._getframe().f_locals['x'] = 2
         print(x)
 
-prints ``2``
-
-but::
+prints ``2``, but::
 
     def f():
         x = 1
         sys._getframe().f_locals['x'] = 2
         print(x)
     f()
 
-prints ``1``
+prints ``1``.
 
-This is inconsistent, and confusing.
-With this PEP both examples would print ``2``.
+This is inconsistent, and confusing. Worse than that, the current behavior can
+result in strange `bugs <https://github.com/python/cpython/issues/74929>`__.
 
-Worse than that, the current behavior can result in strange `bugs
-<https://github.com/python/cpython/issues/74929>`__.
+With this PEP both examples would print ``2`` as the function level
+change would be written directly to the optimized local variables in
+the frame rather than to a cached dictionary snapshot.
 
 There are no compensating advantages for the current behavior;
 it is unreliable and slow.
 
+The ``locals()`` builtin has its own undesirable behaviours, where
+adding a new local variable can change the behaviour of code
+executed with ``exec()`` in function scopes, even if that code
+runs *before* the local variable is defined.
+
+For example::
+
+    def f():
+        exec("x = 1")
+        print(locals().get("x"))
+    f()
+
+prints ``1``, but::
+
+    def f():
+        exec("x = 1")
+        print(locals().get("x"))
+        x = 0
+    f()
+
+prints ``None`` (the default value from the ``.get()`` call).
+
+With this PEP both examples would print ``None``, as the call to
+``exec()`` and the subsequent call to ``locals()`` would use
+independent dictionary snapshots of the local variables rather
+than using the same shared dictionary cached on the frame object.
+
 Rationale
 =========
 
+Making the ``frame.f_locals`` attribute a write-through proxy
+-------------------------------------------------------------
+
 The current implementation of ``frame.f_locals`` returns a dictionary
-that is created on the fly from the array of local variables.
+that is created on the fly from the array of local variables. The
+``PyFrame_LocalsToFast()`` C API is then called by debuggers and trace
+functions that want to write their changes back to the array (until
+Python 3.11, this API was called implicitly after every trace function
+invocation rather than being called explicitly by the trace functions).
+
 This can result in the array and dictionary getting out of sync with
-each other. Writes to the ``f_locals`` may not show up as
-modifications to local variables. Writes to local variables can get lost.
+each other. Writes to the ``f_locals`` frame atribute may not show up as
+modifications to local variables if ``PyFrame_LocalsToFast()`` is never
+called. Writes to local variables can get lost if the dictionary snapshot
+is not refreshed before being written back to the frame.
```

Comment:
I was a bit confused here. I thought writes to local variables get lost if the snapshot *is* refreshed before being written back to the frame. So if you do `exec("x = 1")` then `locals()` (to refresh the snapshot), your `x=1` is lost?